### ISSUE TITLE
CPP_STD CAS kernels giving SEGFAULT

### ISSUE THREAD

**Issue by pbbdasilva** [tactcomplabs/circustent#20]
Have not taken the time to analyze why, but all CAS kernels (from CPP_STD) except the PTRCHASE and CENTRAL are giving me segfault for the following execution

```
$EXE --bench XXX_CAS -m 1030560875 -p 4 -i 10000000
```

I don't know if this is a known issue but thought it was worth bringing it up
Below are 2 screenshots portraying what I just said:

![Screenshot from 2022-10-07 18-19-05](https://user-images.githubusercontent.com/61165929/194674673-5613a195-d882-4e86-aca9-147a9430d339.png)
![Screenshot from 2022-10-07 18-17-27](https://user-images.githubusercontent.com/61165929/194674681-b20aecab-7deb-4a36-90cd-dd4bbba80603.png)

**EDIT**
**CPP_STD**
- cmake -DENABLE_CPP_STD=ON ../
- gcc version 8.3.1 20190507 (Red Hat 8.3.1-4) (GCC)
I am not sure how much memory is because I did not allocate a full node for the run.

**Comment by pbbdasilva** [comment]
![image](https://user-images.githubusercontent.com/61165929/194728818-e902e898-6be5-47f8-a17c-f4dbc652a7c7.png)
Seems to be a stack overflow problem
allocating dynamically would fix but may decrease performance
```cpp
uint64_t *expected;
expected = (uint64_t *) malloc(sizeof(uint64_t) * iters);
```

**Comment by jleidel** [comment]
@pbbdasilva I believe these are two different bugs. First, for each issue, please note the following info in the issue:

1. The cmake compilation options (which backends where enabled)
2. The compiler utilized (`gcc --version`; `clang --version`; etc)
3. The target system details (processor, num cores, gpu, memory capacity, etc)

This will give everyone a better idea of "who" should be assigned the bug and "where" to look for the potential issue.  I tested the same command line options on the following configuration(s): 

**OMP**
- cmake -DENABLE_OMP=ON
- gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
- ubuntu 20.04; 20 physical cores (40 virtual); 128GB memory

**CPP_STD**
- cmake -DENABLE_CPP_STD=ON
- gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
- ubuntu 20.04; 20 physical cores (40 virtual); 128GB memory

I used the following runtime option(s): `./src/CircusTent/circustent -b RAND_CAS -m 1030560875 -p 4 -i 10000000`

The **OMP** kernel passes without issue.  The **CPP_STD** kernel fails with a seg fault.  I suspect this bug is directly attributed to the **CPP_STD** implementation and not all backends.  Compiling with debug flags and pulling a full trace gives us:

```
0x000055555555cd84 in CT_CPP_STD::RAND_CAS (this=0x5555555782b0, thread_id=0,
    barrier_ctr=0x7fffffffe0f8, start_time=0x7fffffffe0f0)
    at /Users/jleidel/dev/working/circustent/src/CircusTent/Impl/CT_CPP_STD/CT_CPP_STD_IMPL.cpp:43
43	  uint64_t expected[iters];
(gdb) bt
#0  0x000055555555cd84 in CT_CPP_STD::RAND_CAS (this=0x5555555782b0, thread_id=0,
    barrier_ctr=0x7fffffffe0f8, start_time=0x7fffffffe0f0)
    at /Users/jleidel/dev/working/circustent/src/CircusTent/Impl/CT_CPP_STD/CT_CPP_STD_IMPL.cpp:43
#1  0x000055555555cb45 in std::__invoke_impl<void, void (CT_CPP_STD::*)(unsigned long, std::atomic<unsigned long>*, double*), CT_CPP_STD*, unsigned long, std::atomic<unsigned long>*, double*> (__f=
    @0x555555578758: (void (CT_CPP_STD::*)(CT_CPP_STD * const, unsigned long, std::atomic<unsigned long> *, double *)) 0x55555555ccc4 <CT_CPP_STD::RAND_CAS(unsigned long, std::atomic<unsigned long>*, double*)>, __t=@0x555555578750: 0x5555555782b0) at /usr/include/c++/9/bits/invoke.h:73
#2  0x000055555555c97f in std::__invoke<void (CT_CPP_STD::*)(unsigned long, std::atomic<unsigned long>*, double*), CT_CPP_STD*, unsigned long, std::atomic<unsigned long>*, double*> (__fn=
    @0x555555578758: (void (CT_CPP_STD::*)(CT_CPP_STD * const, unsigned long, std::atomic<unsigned long> *, double *)) 0x55555555ccc4 <CT_CPP_STD::RAND_CAS(unsigned long, std::atomic<unsigned long>*, double*)>) at /usr/include/c++/9/bits/invoke.h:95
#3  0x000055555555c80f in std::thread::_Invoker<std::tuple<void (CT_CPP_STD::*)(unsigned long, std::atomic<unsigned long>*, double*), CT_CPP_STD*, unsigned long, std::atomic<unsigned long>*, double*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul> (this=0x555555578738) at /usr/include/c++/9/thread:244
#4  0x000055555555c75d in std::thread::_Invoker<std::tuple<void (CT_CPP_STD::*)(unsigned long, std::atomic<unsigned long>*, double*), CT_CPP_STD*, unsigned long, std::atomic<unsigned long>*, double*> >::operator() (this=0x555555578738) at /usr/include/c++/9/thread:251
#5  0x000055555555c72e in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (CT_CPP_STD::*)(unsigned long, std::atomic<unsigned long>*, double*), CT_CPP_STD*, unsigned long, std::atomic<unsigned long>*, double*> > >::_M_run (this=0x555555578730) at /usr/include/c++/9/thread:195
#6  0x00007ffff7ea7de4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#7  0x00007ffff7d9b609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#8  0x00007ffff7cc0133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Regarding your `malloc` note, I would avoid that route in favor of utilizing dynamically allocated arrays in C++.  
```
uint64_t *expected = new uint64_t[iters];
...
delete[] expected;
```

**Comment by pbbdasilva** [comment]
Hey @jleidel thanks, I will start following what you said about the info I should provide.
I think I may have expressed myself badly (probably due to not explaining what and how I was running as you pointed out, I am sorry again)
The screenshots are all related to the CPP_STD backend. Some CAS kernels are giving segfault with the number of allocated memory I used in my last comment. My 2nd comment is when I ran the CPP_STD RAND_CAS with `gdb` and I saw that the segfault occurred when allocating memory. That is the reason I thought it was giving us SEGFAULT and why I commented about allocating it dynamically (because with the malloc the kernel was working). However, seeing your backtrace I am not certain anymore and will take a better look into it.
